commit 73f5249f3f14886831a9da5fb7731b0c227b8d69
Author: Adam C. Emerson <aemerson@redhat.com>
Date:   Tue Aug 7 16:40:57 2018 -0400

    monc: Asifoact MonClient
    
    Of course now everyone has to feed an io_context into the MonClient.
    
    Signed-off-by: Adam C. Emerson <aemerson@redhat.com>

diff --git a/src/mgr/MgrStandby.cc b/src/mgr/MgrStandby.cc
index 7d193b6575b..f751b83ee2c 100644
--- a/src/mgr/MgrStandby.cc
+++ b/src/mgr/MgrStandby.cc
@@ -38,7 +38,7 @@
 
 MgrStandby::MgrStandby(int argc, const char **argv) :
   Dispatcher(g_ceph_context),
-  monc{g_ceph_context},
+  monc{g_ceph_context, poolctx},
   client_messenger(Messenger::create(
 		     g_ceph_context,
 		     cct->_conf.get_val<std::string>("ms_type"),
@@ -115,6 +115,8 @@ int MgrStandby::init()
   client_messenger->add_dispatcher_tail(&client);
   client_messenger->start();
 
+  poolctx.start(2);
+
   // Initialize MonClient
   if (monc.build_initial_monmap() < 0) {
     client_messenger->shutdown();
@@ -169,6 +171,7 @@ int MgrStandby::init()
   // this method.
   monc.set_passthrough_monmap();
 
+  poolctx.start(1);
   client_t whoami = monc.get_global_id();
   client_messenger->set_myname(entity_name_t::MGR(whoami.v));
   monc.set_log_client(&log_client);
@@ -287,7 +290,20 @@ void MgrStandby::shutdown()
     }
 
     py_module_registry.shutdown();
-
+    // stop sending beacon first, i use monc to talk with monitors
+    timer.shutdown();
+    // client uses monc and objecter
+    client.shutdown();
+    mgrc.shutdown();
+    // Stop asio threads, so leftover events won't call into shut down
+    // monclient/objecter.
+    poolctx.finish();
+    // stop monc, so mon won't be able to instruct me to shutdown/activate after
+    // the active_mgr is stopped
+    monc.shutdown();
+    if (active_mgr) {
+      active_mgr->shutdown();
+    }
     // objecter is used by monc and active_mgr
     objecter.shutdown();
     // client_messenger is used by all of them, so stop it in the end
@@ -298,6 +314,7 @@ void MgrStandby::shutdown()
   // to touch references to the things we're about to tear down
   finisher.wait_for_empty();
   finisher.stop();
+  poolctx.finish();
 }
 
 void MgrStandby::respawn()
diff --git a/src/mgr/MgrStandby.h b/src/mgr/MgrStandby.h
index 98b9d0d22a7..09ce8428b1a 100644
--- a/src/mgr/MgrStandby.h
+++ b/src/mgr/MgrStandby.h
@@ -16,6 +16,7 @@
 #define MGR_STANDBY_H_
 
 #include "auth/Auth.h"
+#include "common/async/context_pool.h"
 #include "common/Finisher.h"
 #include "common/Timer.h"
 #include "common/LogClient.h"
@@ -39,6 +40,7 @@ public:
 			  const std::set <std::string> &changed) override;
 
 protected:
+  ceph::async::io_context_pool poolctx;
   MonClient monc;
   std::unique_ptr<Messenger> client_messenger;
   Objecter objecter;
